### PR TITLE
refactor(parser) : move Materials from potree parsers to potree provider (+custom layer material option)

### PR DESCRIPTION
**Pull request by mbredif**
## Description
Potree parsers used a hardcoded itowns.PointsMaterial  material
This PR tackles some part of #695 to decouple the parsing and the styling of potree pointclouds.

## Change
- Potree parsers now output a THREE.BufferGeometry, rather than a THREE.Points.
- The potree PointCloudProvider (to be renamed more explicitly as well as the pointcloud examples by the way :) ) is now responsible for creating the THREE.Points and adding the layer material (which defaults to itowns.PointsMaterial : Potree Parser now have no knowledge of materials and rendering logic.
- customBinFormat is now a local variable (private implementation detail)
- realPointCount is removed as it was not used apparently (if needed, it can be accessed using the size of the position attribute)
- the pointcloud.html example shows a sample usage of potree layer with a custom material : [a simple upstream THREE.PointsMaterial](https://github.com/iTowns/itowns/compare/potree_custom_material?expand=1#diff-d619f5f9635467ba864c74d37cfcda7fR185).

## Screenshots
- http://localhost:8080/examples/pointcloud.html : No visible change
- http://localhost:8080/examples/pointcloud.html?material=three
image
(note the red diffuse color and the square sprites)

## Note
This PR is a slight rework of the first commit of #760. The second commit that allows sharing the material with no cloning will be reworked in its own PR.

### CODE REVIEW

**Comment by mbredif** [comment]
Rebased and ready to review ! (after #764)

File: src/Provider/PointCloudProvider.js
Comment (peppsac) [inline src/Provider/PointCloudProvider.js]: Just a comment here: I'd prefer to keep the old mechanism (update during process).

That way the `onBeforeRender` callbacks provided by `three` are still available for the user.

**Comment by mbredif** [comment]
@autra any comment on this ? (it is complementary with #767 )

File: src/Provider/PointCloudProvider.js
Comment (autra) [reply src/Provider/PointCloudProvider.js]: +1 :-)

File: src/Provider/PointCloudProvider.js
Comment (autra) [inline src/Provider/PointCloudProvider.js]: points.tightbbox is actually never used directly, but copied into elt.tightbbox in processing. As we now use geometry.boundingBox, maybe we can take this opportunity to remove an intermediary step and store geometry.boundingBox directly into elt.tightbbox in processing? Also it'll allow elt.tightbbox be always up-to-date if we always use `geometry.computeBoundingBox()` after that.

File: src/Renderer/PointsMaterial.js
Comment (autra) [inline src/Renderer/PointsMaterial.js]: not a big deal, but why did you change `update...`  to `refresh...` ?

File: examples/pointcloud.html
Comment (autra) [inline examples/pointcloud.html]: nits: `new URL(...).searchParams` is used 2 times, let's define a `var searchParams` on top.

**Review by autra — CHANGES_REQUESTED** [review]
> it is complementary with #767

Not really? Or I didn't see the link apart the fact both touches pointclouds...

But anyway, this looks good to me overall, apart the `onBeforeRender` thing.

It makes sense to return a three.js object here (and not to return a Feature in this case), because atm we use feature only to have either mesh or raster, and I don't see a use case for pointclouds in raster form (yet).

File: src/Provider/PointCloudProvider.js
Comment (peppsac) [inline src/Provider/PointCloudProvider.js]: `points.material.uniforms.overlayColor` should be set as well. Can you add this?

Comment (mbredif) [reply src/Provider/PointCloudProvider.js]: this is now done by the inside the PointsMaterial copy function (I dont want anything material specific such as overlayColor in the provider!)

File: examples/pointcloud.html
Comment (mbredif) [reply examples/pointcloud.html]: done (9529b2c)

File: src/Renderer/PointsMaterial.js
Comment (mbredif) [reply src/Renderer/PointsMaterial.js]: changed back to `updateUniforms` (I was following the naming pattern in three for other onbeforerender callbacks)

File: src/Provider/PointCloudProvider.js
Comment (mbredif) [reply src/Provider/PointCloudProvider.js]: I did not want to change to much in the `PointCloudProcess`, so I only simplified the tightbox computation to a one liner in the  `PointCloudProvider`. https://github.com/iTowns/itowns/blob/ab92777fdaeaaf5a488772701e2156858c0256d0/src/Provider/PointCloudProvider.js#L248

**Comment by mbredif** [comment]
> Not really? Or I didn't see the link apart the fact both touches pointclouds...

 #767 makes all attributes available, this PR lets you provide a custom material to the potree layer, so that client code may design any fancy styling of all these attributes.

> But anyway, this looks good to me overall, apart the onBeforeRender thing.

onBeforeRender is now untouched.

@autra, WDYT ?

File: src/Process/PointCloudProcessing.js
Comment (mbredif) [inline src/Process/PointCloudProcessing.js:167]: this implies `elt.obj.material.visible = true;` if the `layer.material` is visible

File: src/Provider/PointCloudProvider.js
Comment (mbredif) [inline src/Provider/PointCloudProvider.js:266]: material update is now triggered in the Process update (previously transparency was not updated !)

**Comment by mbredif** [comment]
updated and ready for review ! (testing is fixed)

File: utils/debug/PointCloudDebug.js
Comment (autra) [inline utils/debug/PointCloudDebug.js]: The same checkbox is called `Visible` for other layers. It makes sense because it also hides the bounding box btw. 

Also for me it's not a matter of styling but more something useful to hide temporarily a layer. Can you move this on the parent folder?

**Review by autra — COMMENTED** [review]
LGTM, just a little cosmetic change on the debug tools and we can merge that

**Comment by autra** [comment]
Also can you resolve the conflicts and squash your commits?

**Comment by mbredif** [comment]
resolved and squashed !

File: utils/debug/PointCloudDebug.js
Comment (autra) [inline utils/debug/PointCloudDebug.js]: I'd really move it in the parent folder, it's very useful to be able to hide / show layers without having to unfold something, WDYT?

Comment (mbredif) [reply utils/debug/PointCloudDebug.js]: no strong opinion. I have just amended the commit to have the following UI layout :

image

**Review by autra — APPROVED** [review]
Thanks!

**Comment by autra** [comment]
Actually, it seems that the case where point size is 0 is broken... Could you have a look please?

**Review by autra — COMMENTED** [review]
Dismissing my review to avoid accidental merge.

**Review by autra — CHANGES_REQUESTED** [review]
(seems I can't dismiss :-( )

**Comment by mbredif** [comment]
There is no automatic sizing with the THREE material so testing http://localhost:8080/examples/pointcloud.html?material=three with size=0 will give you 0-sized points (hence nothing to see). Is that what you are talking about ?

**Review by autra — APPROVED** [review]

**Comment by autra** [comment]
Oh right, silly me! Sorry for the noise :-)

LGTM, thanks!

File: utils/debug/PointCloudDebug.js
Comment (peppsac) [inline utils/debug/PointCloudDebug.js:16]: I missed this before the merge, but this breaks other parts of itowns which expect `layer.visible`:
at the very least the test here https://github.com/iTowns/itowns/blob/master/src/Core/MainLoop.js#L111 is broken :(

It also has a different behavior than `3dtiles` layer which already supported having a layer.material property: in the 3dtiles case the material is cloned and layer.material is never copied in object.material.

Btw, https://github.com/iTowns/itowns/pull/764/files#diff-6b60bae2240ab556c955b4eb0a07a12eR167 is a bad idea as well IMHO: a user can't set one parameter for a given point object material because it will always be overriden with the one from the layer (so the overlay color feature can now only be enabled globally).

(edit: fixed the 2nd link)

File: src/Renderer/PointsMaterial.js
Comment (peppsac) [inline src/Renderer/PointsMaterial.js:54]: Another too late comment: you're overriding three.js `copy` method but you don't call the `super` method.
This should be fixed:
- either rename your `copy` method and keep its current code
- keep the `copy` name and call RawShaderMaterial.copy as well

**Comment by autra** [comment]
@mbredif Sorry for not catching these problems during my review :-/

I just restored the branch and I'm going to revert this commit for now, if that's ok for you.

**Comment by zarov** [comment]
> I just restored the branch and I'm going to revert this commit for now, if that's ok for you.

Please no, we don't need to revert commits every time we missed something in a review. Another PR will be opened, adding fixes to the remarks that have not been answered. The mistake is human, you validated it, no problem, it happens. If there are bugs, we should open an issue.

**Comment by autra** [comment]
@zarov as you wish :-)

File: utils/debug/PointCloudDebug.js
Comment (mbredif) [reply utils/debug/PointCloudDebug.js:16]: > I missed this before the merge, but this breaks other parts of itowns which expect layer.visible:

For this one a simple fix would be to treat `visible` as `opacity` is treated, with a copy in the `preUpdate`.

Comment (mbredif) [reply utils/debug/PointCloudDebug.js:16]: For the rest of the comment, I left `points.onBeforeRender` untouched just for that use case :).

That way, I think you can still get your per-tile functionality, while fixing the bug present before the merge : if you change styling properties on the layer, then it will only apply to newly created tiles, and not to previously created tiles (just take an earlier commit, change the point size and move around to see that in action).

That being said, maybe we should open an issue to agree and document properly a common strategy for the style/material of all layers.

File: src/Renderer/PointsMaterial.js
Comment (mbredif) [reply src/Renderer/PointsMaterial.js:54]: I have just tried proper overriding, but it does not work (unless you manually set `this.uniforms = source.uniforms` after the call to `super`). I do not know if it is a misuse or an upstream bug, so meanwhile, renaming `copy` to `update` could do the trick.

File: utils/debug/PointCloudDebug.js
Comment (peppsac) [reply utils/debug/PointCloudDebug.js:16]: I agree there was a bug before the merge.

I disagree that `onBeforeRender` is an acceptable substitute tough: having to register a callback for each object you want to customize vs modifying their material once is quite different (looks very much like pushing the complexity on the user of itowns to ease the implementation in itowns core).

That being said, this is a minor feature, so we can discuss this in an issue and try to find a solution that works for all kind of layer.

In the meantime please fix the other issues, especially the `visible`. Thanks!

**Comment by mbredif** [comment]
follow up fixing pr : #780

**Pull request merged**
